## *CSV Input/Output Continued*

In [38]:
## Download CSV metadata on all artists in MoMA's collection

import os
os.chdir('/sharedfolder/')

!wget https://media.githubusercontent.com/media/MuseumofModernArt/collection/master/Artworks.csv

--2018-02-01 19:52:45--  https://media.githubusercontent.com/media/MuseumofModernArt/collection/master/Artists.csv
Resolving media.githubusercontent.com (media.githubusercontent.com)... 151.101.48.133
Connecting to media.githubusercontent.com (media.githubusercontent.com)|151.101.48.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1005208 (982K) [text/plain]
Saving to: ‘Artists.csv.1’

Artists.csv.1       100%[===================>] 981.65K  2.41MB/s    in 0.4s    

2018-02-01 19:52:46 (2.41 MB/s) - ‘Artists.csv.1’ saved [1005208/1005208]

--2018-02-01 19:52:47--  https://media.githubusercontent.com/media/MuseumofModernArt/collection/master/Artworks.csv
Resolving media.githubusercontent.com (media.githubusercontent.com)... 151.101.48.133
Connecting to media.githubusercontent.com (media.githubusercontent.com)|151.101.48.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 55018582 (52M) [text/plain]
Saving to: ‘Artworks.csv.2’

Artwor

In [42]:
## Load 'Artists.csv' as a list of lists called 'table'

import csv

csv_path = "/sharedfolder/Artists.csv"
table = []

with open(csv_path) as file_in:
    csv_input = csv.reader(file_in)
    for row in csv_input:
        table.append(row)

table[:3]

[['\ufeffConstituentID',
  'DisplayName',
  'ArtistBio',
  'Nationality',
  'Gender',
  'BeginDate',
  'EndDate',
  'Wiki QID',
  'ULAN'],
 ['1',
  'Robert Arneson',
  'American, 1930–1992',
  'American',
  'Male',
  '1930',
  '1992',
  '',
  ''],
 ['2',
  'Doroteo Arnaiz',
  'Spanish, born 1936',
  'Spanish',
  'Male',
  '1936',
  '0',
  '',
  '']]

In [43]:
## Removing the column header row and assigning it to its own variable

header_row = table[0]

table = table[1:]

header_row

['\ufeffConstituentID',
 'DisplayName',
 'ArtistBio',
 'Nationality',
 'Gender',
 'BeginDate',
 'EndDate',
 'Wiki QID',
 'ULAN']

In [45]:
## Viewing column headers next to their index values

list(enumerate(header_row))

[(0, '\ufeffConstituentID'),
 (1, 'DisplayName'),
 (2, 'ArtistBio'),
 (3, 'Nationality'),
 (4, 'Gender'),
 (5, 'BeginDate'),
 (6, 'EndDate'),
 (7, 'Wiki QID'),
 (8, 'ULAN')]

In [46]:
## Using list comprehension format to extract all birth date values in the column labeled 'BeginDate'

birth_years = [row[5] for row in table]

birth_years

['1930',
 '1936',
 '1941',
 '1946',
 '1941',
 '1925',
 '1941',
 '1923',
 '1918',
 '1886',
 '1936',
 '1911',
 '1907',
 '1900',
 '0',
 '1923',
 '1926',
 '0',
 '1863',
 '1927',
 '1867',
 '1940',
 '1885',
 '1923',
 '1926',
 '1941',
 '0',
 '0',
 '1898',
 '1894',
 '1932',
 '1921',
 '1930',
 '1883',
 '1888',
 '1898',
 '1913',
 '1900',
 '1937',
 '1934',
 '1933',
 '1935',
 '1928',
 '1920',
 '1933',
 '1940',
 '1915',
 '0',
 '1904',
 '1904',
 '1818',
 '1935',
 '1902',
 '1930',
 '1918',
 '1864',
 '1948',
 '1897',
 '1937',
 '1950',
 '1912',
 '1916',
 '1895',
 '1915',
 '1910',
 '1927',
 '1948',
 '1912',
 '0',
 '1928',
 '1896',
 '1945',
 '1913',
 '1827',
 '1937',
 '1904',
 '1922',
 '1901',
 '1908',
 '1906',
 '1928',
 '1940',
 '1942',
 '1939',
 '1887',
 '1899',
 '1888',
 '0',
 '1943',
 '1897',
 '0',
 '1917',
 '1910',
 '0',
 '1943',
 '1927',
 '1939',
 '0',
 '1928',
 '0',
 '1941',
 '1890',
 '1900',
 '1936',
 '1902',
 '1964',
 '1943',
 '1890',
 '1929',
 '1914',
 '1943',
 '1926',
 '1933',
 '1924',
 '1814'

In [47]:
## Converting each 'BeginDate' value to an integer with a list comprehension

birth_years = [int(row[5]) for row in table]

birth_years

[1930,
 1936,
 1941,
 1946,
 1941,
 1925,
 1941,
 1923,
 1918,
 1886,
 1936,
 1911,
 1907,
 1900,
 0,
 1923,
 1926,
 0,
 1863,
 1927,
 1867,
 1940,
 1885,
 1923,
 1926,
 1941,
 0,
 0,
 1898,
 1894,
 1932,
 1921,
 1930,
 1883,
 1888,
 1898,
 1913,
 1900,
 1937,
 1934,
 1933,
 1935,
 1928,
 1920,
 1933,
 1940,
 1915,
 0,
 1904,
 1904,
 1818,
 1935,
 1902,
 1930,
 1918,
 1864,
 1948,
 1897,
 1937,
 1950,
 1912,
 1916,
 1895,
 1915,
 1910,
 1927,
 1948,
 1912,
 0,
 1928,
 1896,
 1945,
 1913,
 1827,
 1937,
 1904,
 1922,
 1901,
 1908,
 1906,
 1928,
 1940,
 1942,
 1939,
 1887,
 1899,
 1888,
 0,
 1943,
 1897,
 0,
 1917,
 1910,
 0,
 1943,
 1927,
 1939,
 0,
 1928,
 0,
 1941,
 1890,
 1900,
 1936,
 1902,
 1964,
 1943,
 1890,
 1929,
 1914,
 1943,
 1926,
 1933,
 1924,
 1814,
 1925,
 1902,
 1953,
 1939,
 1858,
 0,
 1933,
 1943,
 0,
 1918,
 1936,
 1948,
 1925,
 1868,
 1907,
 0,
 1937,
 1907,
 1933,
 1932,
 1936,
 1919,
 1953,
 1891,
 1903,
 1948,
 1921,
 1813,
 1938,
 1928,
 1933,
 1880,
 1957,
 1955,

In [34]:
## Converting each 'BeginDate' value to an integer, then skipping all zeroes

birth_years = [int(row[5]) for row in table if int(row[5])!=0]

birth_years

[1930,
 1936,
 1941,
 1946,
 1941,
 1925,
 1941,
 1923,
 1918,
 1886,
 1936,
 1911,
 1907,
 1900,
 1923,
 1926,
 1863,
 1927,
 1867,
 1940,
 1885,
 1923,
 1926,
 1941,
 1898,
 1894,
 1932,
 1921,
 1930,
 1883,
 1888,
 1898,
 1913,
 1900,
 1937,
 1934,
 1933,
 1935,
 1928,
 1920,
 1933,
 1940,
 1915,
 1904,
 1904,
 1818,
 1935,
 1902,
 1930,
 1918,
 1864,
 1948,
 1897,
 1937,
 1950,
 1912,
 1916,
 1895,
 1915,
 1910,
 1927,
 1948,
 1912,
 1928,
 1896,
 1945,
 1913,
 1827,
 1937,
 1904,
 1922,
 1901,
 1908,
 1906,
 1928,
 1940,
 1942,
 1939,
 1887,
 1899,
 1888,
 1943,
 1897,
 1917,
 1910,
 1943,
 1927,
 1939,
 1928,
 1941,
 1890,
 1900,
 1936,
 1902,
 1964,
 1943,
 1890,
 1929,
 1914,
 1943,
 1926,
 1933,
 1924,
 1814,
 1925,
 1902,
 1953,
 1939,
 1858,
 1933,
 1943,
 1918,
 1936,
 1948,
 1925,
 1868,
 1907,
 1937,
 1907,
 1933,
 1932,
 1936,
 1919,
 1953,
 1891,
 1903,
 1948,
 1921,
 1813,
 1938,
 1928,
 1933,
 1880,
 1957,
 1955,
 1926,
 1935,
 1942,
 1923,
 1902,
 1888,
 1930,
 1914,

In [48]:
## Calculating the mean birth year with numpy

import numpy as np

np.mean(birth_years)

1454.2549782707401

## *Filtering rows*

In [49]:
## Write a piece of code that creates a new table (i.e., list of lists)
## containing only artists born in the 1880s.


born_1880s = []

for row in table:
    if 1880 <= int(row[5]) <= 1889:
        born_1880s.append(row)


print(len(born_1880s))

539


In [50]:
## View 3 random rows in our 'born_1880s' list of lists

import random

random.sample(born_1880s, 3)

[['5097',
  'Anne Ryan',
  'American, 1889–1954',
  'American',
  'Female',
  '1889',
  '1954',
  'Q4768767',
  '500020957'],
 ['8400',
  'Heinz Fuchs',
  'German, 1886–1961',
  'German',
  'Male',
  '1886',
  '1961',
  '',
  ''],
 ['624',
  'Umberto Boccioni',
  'Italian, 1882–1916',
  'Italian',
  'Male',
  '1882',
  '1916',
  'Q152797',
  '500012624']]

### *Average Artist Age*

Now that we’ve defined a meaningful subset of our data, let’s see what we can do with it. For instance, what was the mean life span of artists born in the 1880s (who happen to be included in MoMA's collections)?

In [51]:
lifespans_1880s = []

for row in born_1880s:
    lifespans_1880s.append(int(row[6]) - int(row[5]))

lifespans_1880s

[80,
 55,
 90,
 65,
 82,
 88,
 76,
 88,
 77,
 78,
 66,
 80,
 89,
 86,
 56,
 71,
 66,
 78,
 65,
 72,
 66,
 66,
 36,
 77,
 86,
 43,
 52,
 86,
 73,
 89,
 88,
 76,
 79,
 78,
 34,
 50,
 50,
 87,
 71,
 67,
 60,
 52,
 57,
 81,
 84,
 85,
 55,
 75,
 74,
 83,
 85,
 68,
 58,
 70,
 85,
 80,
 58,
 71,
 98,
 87,
 62,
 90,
 84,
 74,
 45,
 103,
 84,
 86,
 78,
 75,
 84,
 93,
 64,
 85,
 85,
 98,
 65,
 63,
 63,
 56,
 94,
 52,
 74,
 71,
 55,
 41,
 75,
 66,
 81,
 90,
 62,
 85,
 80,
 79,
 67,
 60,
 83,
 49,
 81,
 77,
 73,
 74,
 77,
 78,
 83,
 82,
 71,
 80,
 84,
 71,
 51,
 53,
 62,
 59,
 -1881,
 60,
 58,
 98,
 74,
 76,
 72,
 59,
 81,
 -1880,
 87,
 91,
 57,
 80,
 93,
 47,
 40,
 94,
 86,
 59,
 70,
 46,
 77,
 75,
 74,
 85,
 64,
 87,
 70,
 87,
 78,
 85,
 89,
 86,
 85,
 96,
 79,
 77,
 76,
 79,
 56,
 73,
 73,
 76,
 83,
 67,
 71,
 58,
 66,
 80,
 95,
 72,
 89,
 47,
 58,
 87,
 66,
 74,
 94,
 78,
 76,
 90,
 83,
 84,
 94,
 64,
 40,
 56,
 53,
 83,
 73,
 69,
 78,
 65,
 80,
 38,
 81,
 75,
 55,
 77,
 80,
 89,
 89,
 81,
 27

If you scroll through your list of lifespans, you’ll see occasional negative numbers (e.g., “-1887”). Since missing values are represented by “0,” if no death date is listed we’ll end up subtracting an artist’s birth year from zero. Let’s amend our code to leave out these rows.


In [57]:
lifespans_1880s = []

for row in born_1880s:
    age = int(row[6])-int(row[5])
    if age > 0:
        lifespans_1880s.append(age)

lifespans_1880s

[80,
 55,
 90,
 65,
 82,
 88,
 76,
 88,
 77,
 78,
 66,
 80,
 89,
 86,
 56,
 71,
 66,
 78,
 65,
 72,
 66,
 66,
 36,
 77,
 86,
 43,
 52,
 86,
 73,
 89,
 88,
 76,
 79,
 78,
 34,
 50,
 50,
 87,
 71,
 67,
 60,
 52,
 57,
 81,
 84,
 85,
 55,
 75,
 74,
 83,
 85,
 68,
 58,
 70,
 85,
 80,
 58,
 71,
 98,
 87,
 62,
 90,
 84,
 74,
 45,
 103,
 84,
 86,
 78,
 75,
 84,
 93,
 64,
 85,
 85,
 98,
 65,
 63,
 63,
 56,
 94,
 52,
 74,
 71,
 55,
 41,
 75,
 66,
 81,
 90,
 62,
 85,
 80,
 79,
 67,
 60,
 83,
 49,
 81,
 77,
 73,
 74,
 77,
 78,
 83,
 82,
 71,
 80,
 84,
 71,
 51,
 53,
 62,
 59,
 60,
 58,
 98,
 74,
 76,
 72,
 59,
 81,
 87,
 91,
 57,
 80,
 93,
 47,
 40,
 94,
 86,
 59,
 70,
 46,
 77,
 75,
 74,
 85,
 64,
 87,
 70,
 87,
 78,
 85,
 89,
 86,
 85,
 96,
 79,
 77,
 76,
 79,
 56,
 73,
 73,
 76,
 83,
 67,
 71,
 58,
 66,
 80,
 95,
 72,
 89,
 47,
 58,
 87,
 66,
 74,
 94,
 78,
 76,
 90,
 83,
 84,
 94,
 64,
 40,
 56,
 53,
 83,
 73,
 69,
 78,
 65,
 80,
 38,
 81,
 75,
 55,
 77,
 80,
 89,
 89,
 81,
 27,
 72,
 83,
 65,

Now that we have a list of valid integers, all we need to do is calculate the mean. Below we divide the sum of the list (which we cast as a float) by its length to get 72.64 years.


In [59]:
## Mean lifespan of artists born in the 1880s

float(sum(lifespans_1880s)) / len(lifespans_1880s)

72.74269005847954

In [60]:
## Mean lifespan of artists born in the 1880s
# A common convention is to rename the NumPy package to 'np' at the import step.

import numpy as np

np.mean(lifespans_1880s)

72.74269005847954